stellar_uri.go

```diff
@@ -0,0 +1,336 @@
+
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+	if u.Scheme != "web+stellar" {
+		return nil, ErrInvalidScheme
+	}
+
+	res.Operation = u.Opaque
+
+	res.values = u.Query()
+
+	res.OriginDomain = res.value("origin_domain")
+	res.Signature = res.value("signature")
+
+	return res, nil
+
+}
+
+func (u *unvalidatedURI) Validate(getter HTTPGetter) (*ValidatedStellarURI, error) {
+	// origin_domain and signature are optional in the spec, but
+	// it seems like a really bad idea to allow any of these
+	// requests without them, so we are going to make them required.
+	if u.OriginDomain == "" {
+		return nil, ErrMissingParameter{Key: "origin_domain"}
+	}
+	if u.Signature == "" {
+		return nil, ErrMissingParameter{Key: "signature"}
+	}
+
+	// make sure there's no port or scheme
+	if strings.IndexByte(u.OriginDomain, ':') != -1 {
+		return nil, ErrInvalidParameter{Key: "origin_domain"}
+	}
+
+	switch u.Operation {
+	case "pay":
+		return u.validatePay(getter)
+	case "tx":
+		return u.validateTx(getter)
+	default:
+		return nil, ErrInvalidOperation
+	}
+}
+
+type tomlStellar struct {
+	SigningKey string `toml:"URI_REQUEST_SIGNING_KEY"`
+}
+
+func (u *unvalidatedURI) originDomainSigningKey(getter HTTPGetter) (string, error) {
+	wellKnownURL := fmt.Sprintf("https://%s/.well-known/stellar.toml", u.OriginDomain)
+	_, err := url.Parse(wellKnownURL)
+	if err != nil {
+		return "", ErrInvalidParameter{Key: "origin_domain"}
+	}
+
+	resp, err := getter.Get(wellKnownURL)
+	if err != nil {
+		return "", ErrNetworkWellKnownOrigin{Wrapped: err}
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", ErrInvalidWellKnownOrigin{Wrapped: err}
+	}
+	if resp.StatusCode != http.StatusOK {
+		return "", ErrInvalidWellKnownOrigin{Wrapped: errors.New("stellar.toml not found")}
+	}
+
+	var sdoc tomlStellar
+	if _, err := toml.Decode(string(body), &sdoc); err != nil {
+		return "", ErrInvalidWellKnownOrigin{Wrapped: err}
+	}
+
+	return strings.TrimSpace(sdoc.SigningKey), nil
+}
+
+func (u *unvalidatedURI) validateOriginDomain(getter HTTPGetter) error {
+	signingKey, err := u.originDomainSigningKey(getter)
+	if err != nil {
+		return err
+	}
+
+	if signingKey == "" {
+		return ErrInvalidWellKnownOrigin{Wrapped: errors.New("no signing key")}
+	}
+
+	kp, err := keypair.Parse(signingKey)
+	if err != nil {
+		return ErrInvalidWellKnownOrigin{Wrapped: errors.New("invalid signing key")}
+	}
+
+	signature, err := base64.StdEncoding.DecodeString(u.Signature)
+	if err != nil {
+		return ErrBadSignature
+	}
+
+	if err := kp.Verify(u.payload(), signature); err != nil {
+		return ErrBadSignature
+	}
+
+	return nil
+}
+
+func (u *unvalidatedURI) payload() []byte {
+	// get the portion of the URI that was signed by stripping &signature off the end
+	index := strings.LastIndex(u.raw, "&signature=")
+	if index == -1 {
+		// this shouldn't happen because we already checked that signature
+		// exists
+		return nil
+	}
```

Comment:
would return a subset of the parameters if signature isn't the last query parameter

Reply:
The spec says that signature should be the last parameter.  If they don't do that, then can't make the payload and the signature will fail.  So if they somehow put the signature somewhere else, that should fail, and it will...

Reply:
Take the uri
```
web+stellar:pay?destination=GCALNQQBXAPZ2WIRSDDBMSTAKCUH5SG6U76YBFLQLIXJTF7FE5AX7AOO&amount=120.1234567&signature=xsomesigx&asset_code=USD&asset_issuer=GSOMEISSUER
```
Where `xsomesigx` is a sig over
```
web+stellar:pay?destination=GCALNQQBXAPZ2WIRSDDBMSTAKCUH5SG6U76YBFLQLIXJTF7FE5AX7AOO&amount=120.1234567
```
Maybe I messed up the details but the idea is that the signature could be only over the parameters that come before `signature`. And then a party not in control of the signing key could append additional parameters.

That doesn't conform to the spec but I think this signature checking would OK it.

Reply:
Ok, I changed it and added a test for a parameter after the signature.